grammar/cypher.xml

```diff
@@ -89,22 +85,107 @@
     </alt>
   </production>
 
-  <production name="Clause" rr:inline="true">
+  <production name="SingleQuery" rr:inline="true">
     <alt>
-      <non-terminal ref="LoadCSV"/>
-      <non-terminal ref="Start"/>
-      <non-terminal ref="Match"/>
-      <non-terminal ref="Unwind"/>
+      <non-terminal ref="SinglePartQuery"/>
+      <non-terminal ref="MultiPartQuery"/>
+    </alt>
+  </production>
+
+  <production name="SinglePartQuery">
+    <alt>
+      <non-terminal ref="ReadOnlyEnd"/>
+      <non-terminal ref="ReadUpdateEnd"/>
+      <non-terminal ref="UpdatingEnd"/>
+    </alt>
+  </production>
+
+  <production name="ReadOnlyEnd" rr:inline="true">
+    <non-terminal ref="ReadPart"/>
+    <non-terminal ref="Return"/>
+  </production>
+
+  <production name="ReadUpdateEnd" rr:inline="true">
+    <non-terminal ref="ReadingClauseExceptWith"/>
+    <repeat>
+      &WS; <non-terminal ref="ReadingClauseExceptWith"/>
+    </repeat>
+    <repeat min="1">
+      &WS; <non-terminal ref="UpdatingClause"/>
+    </repeat>
+    <opt>&WS; <non-terminal ref="Return"/></opt>
+  </production>
+
+  <production name="UpdatingEnd" rr:inline="true">
+    <non-terminal ref="UpdatingStartClause"/>
+    <repeat>
+      &WS; <non-terminal ref="UpdatingClause"/>
+    </repeat>
+    <opt>&WS; <non-terminal ref="Return"/></opt>
+  </production>
+
+  <production name="MultiPartQuery">
+    <alt>
+      <non-terminal ref="ReadPart"/>
+      <seq>
+        <non-terminal ref="UpdatingStartClause"/>
+        &WS;
+        <non-terminal ref="UpdatingPart"/>
+      </seq>
+    </alt>
+
+    <non-terminal ref="With"/>
+    &WS;
+
+    <repeat>
+      <non-terminal ref="ReadPart"/>
+      <non-terminal ref="UpdatingPart"/>
+      <non-terminal ref="With"/> &WS;
+    </repeat>
+
+    <non-terminal ref="SinglePartQuery"/>
+  </production>
+
+  <production name="ReadPart" rr:inline="true">
+    <repeat>
+      <non-terminal ref="ReadingClauseExceptWith"/> &WS;
+    </repeat>
+  </production>
+
+  <production name="UpdatingPart" rr:inline="true">
+    <repeat>
+      <non-terminal ref="UpdatingClause"/> &WS;
+    </repeat>
+  </production>
+
+  <production name="UpdatingStartClause" rr:inline="true">
+    <alt>
+      <non-terminal ref="Create"/>
       <non-terminal ref="Merge"/>
+      <non-terminal ref="CreateUnique"/>
+      <non-terminal ref="Foreach"/>
+    </alt>
+  </production>
+
+  <production name="UpdatingClause" rr:inline="true">
+    <alt>
       <non-terminal ref="Create"/>
+      <non-terminal ref="Merge"/>
       <non-terminal ref="CreateUnique"/>
-      <non-terminal ref="Set"/>
+      <non-terminal ref="Foreach"/>
       <non-terminal ref="Delete"/>
+      <non-terminal ref="Set"/>
       <non-terminal ref="Remove"/>
-      <non-terminal ref="Foreach"/>
+    </alt>
+  </production>
+
+  <production name="ReadingClauseExceptWith" rr:inline="true">
```

Comment:
I don't think `ExceptWith` adds anything useful to the name, arguably `WITH` doesn't do any reading anyhow ;)